### ISSUE TITLE
Sorry, I borked steps/combine_ali_dirs.sh

### ISSUE THREAD

**Issue by kkm000** [kaldi-asr/kaldi#3537]
https://github.com/kaldi-asr/kaldi/blob/93bdc41a31fc451ee0a05bd1b1add3397e8dc969/egs/wsj/s5/steps/combine_ali_dirs.sh#L161-L165

If the command `$cmd` literally replaces the tag `JOB` with the job index for each indexed job separately, like `run.pl` does, this works. If, however, the program inserts a cluster scheduler's variable instead, like `slurm.pl`, this results in an unpalatable construct `gunzip -c ${src_arks_${SLURM_ARRAY_TASK_ID}}`.

This is not super-urgent, as, afaik, the step is used by only one eg, and this can be worked around by running it under run.pl.

----

We generally lack a way to pass sets of data indexable by job index to the `$cmd` pl scripts. This has already been noted in the nnet3 train script comments, where each job in the array must receive multiple random cegs indices. I'll see if I can come up with some generalization that would not break the existing run command syntax and be readable at the same time.

**Comment by danpovey** [comment]
Would probably be easiest to just launch each one as a single separate background job, and wait for them.  It wouldn't respect the --max-jobs-run flag, but that shouldn't be deadly

**Comment by kkm000** [comment]
Yes, there are certainly a few possible workarounds. Too bad bash 3.2 does not have `wait -n` (wait for any one job and return its exit status).  I have this function in my common script include file, which waits for any number of background jobs and checks their statuses, but this won't cut it for bash 3.2 (will MacOS ever upgrade their bash).
```
# Wait for all background jobs and check their status.
WaitParallel() {
  # 1. 'jobs -p' does not stop tracking completed jobs.
  # 2. We only scan number of words (PIDs) in its returned value, not
  #    the values, so the 'word' is unused in the loop.
  # 3. Cannot use let when 'set -e' is in effect: let returns its result
  #    as exit code; must use x=$((x+1)) to increment.
  local jobs=$(jobs -p) word nfail=0 ntot=0
  for word in $jobs; do
    ntot=$((ntot + 1))
    wait -n || nfail=$((nfail + 1))
  done
  (( nfail > 0 )) &&
    Die "$nfail out of $ntot parallel jobs failed. Check output above."
  true
}
```
The python nnet3 training scripts do not suffer from this shortcoming--the main program can check the exit code of every background process--but they are not as readable as bash programs. But the 1,2,3 in the comments above number the rakes I've stepped on. Name a scripting language that does not suck...

Let me sit on this for a couple days. If we were living in Narnia, we could be able to pass different arguments to each array index job. But even bash arrays are of no help, as Bash does not allow exporting array values.

**Comment by danpovey** [comment]
I was thinking of just doing
```
for n in $(seq 1 $num_jobs); do
   launch job &
done

wait
```
which waits for all of them.

**Comment by kkm000** [comment]
I was thinking now and then about a generic way of passing arrays to $cmd scripts. How would you rate usefulness/importance of having such a capability? There is also this comment:
https://github.com/kaldi-asr/kaldi/blob/fdab49d9439dceac72b20131a8e14eaf2372a393/egs/wsj/s5/steps/libs/nnet3/train/chain_objf/acoustic_model.py#L137-L141

> We cannot easily use a single parallel SGE job to do the main training,
> because the computation of which archive and which --frame option
> to use for each job is a little complex, so we spawn each one separately.
> this is no longer true for RNNs as we use do not use the --frame option
> but we use the same script for consistency with FF-DNN code

One not too ugly way I can think of is using bash `declare -p`, which prints variables as bash code, in a switch to $cmd, like 
```
$cmd --vars=$(declare -p foo bar) ....
```
and then the value of this switch can be added literally to the generated job scheduler script in e.g. queue.pl.

Indexed arrays work even in bash 3.2. Just to illustrate,
```
bash-3.2# foo=(word 'or whole phrase')
bash-3.2# foo[5]="Pol O'Toole"
bash-3.2# bar=(0 2 4 8)
bash-3.2# vars=$(declare -p foo bar)
bash-3.2# echo "$vars" >script
bash-3.2# cat script
declare -a foo='([0]="word" [1]="or whole phrase" [5]="Pol O'\''Toole")'
declare -a bar='([0]="0" [1]="2" [2]="4" [3]="8")'
bash-3.2# unset foo bar
bash-3.2# echo ${foo[*]}

bash-3.2# . script
bash-3.2# echo ${foo[*]}
word or whole phrase Pol O'Toole
bash-3.2# declare -p foo
declare -a foo='([0]="word" [1]="or whole phrase" [5]="Pol O'\''Toole")'
```
If you do not think that this capability is important, in this script I can use `wait`, just like you suggested, and the `|| touch .error` trick in background subshells, used in a few other places. The general thing is more complex to implement, but not too complex. So what do you think.

**Comment by danpovey** [comment]
For now I think it's best to just do wait.  I'm avoiding major changes of
this type to mainline Kaldi at this point as I want to focus on next-gen
stuff.

**Comment by kkm000** [comment]
Roger.

BTW, I have no clear idea what is the architectural idea of kaldi10 is going to be. Say, for the current Kaldi, I'd put it in one sentence as "a set of executable binaries with common invocation interface, and I/O formats and machinery, with bash as the high-level glue language."  Maybe not “modern” in any sense, but shells are unbeatable as a glue in such a case. Will Kaldi-next be a set of python-friendly API with Python as the glue, or is it deeper than that? How deep a symbiosis is it going to develop with PyTorch¹? I tried to follow, but what I can't _really_ grok is kind of the highest-level central philosophy of the new design. In other words, what I'm missing is that one-sentence description.

___
¹ I bet that by the time people forget what PyTorch had been, bourne-style shells and C++ will still be around.

**Comment by danpovey** [comment]
@kkm I'm actually still trying to figure all that out.  It's not easy.  It
may have to develop over time-- e.g. start by implementing something, see
what issues come up, and be willing to change the design as we go.

**Comment by kkm000** [comment]
I see. I think it's time for me to dig PyTorch deeper. I've been using Keras for ML stuff, because it's backed by TF, and TF models are relatively easy to productize (thanks to it's dual Python and C++ API). And since it works I do not try to fix it--I haven't looked much into other Python-driven frameworks.

**Comment by danpovey** [comment]
It is not 100% set in stone to use PyTorch and not TensorFlow.

**Comment by danpovey** [comment]
Resolved in #3561

**Comment by marvin-nj** [comment]
Hello dan, I am confused that why we do not combine "trans.job" files in each src dir  , which  would be used to do fmllr transform  at the stage of build-tree .

**Comment by danpovey** [comment]
Hm.  Yes, that is an issue.  I forgot that the fMLLR transforms make a difference to the tree building.  There might be some advantage in modifying that script to also re-split the trans files.  If you want you can create a new issue, or create a PR.

**Comment by marvin-nj** [comment]
Okey , I create a new question now， and try to fix it .  thank you!
